csharp/src/Ice/Communicator.cs

```diff
@@ -1193,37 +1193,49 @@ public Timer Timer()
         }
 
         // Registers an endpoint factory.
-        public void IceAddEndpointFactory(IEndpointFactory factory)
+        public void IceAddEndpointFactory(Transport transport, string transportName, IEndpointFactory factory)
         {
-            _typeToEndpointFactory.Add(factory.Type, factory);
-            _transportToEndpointFactory.Add(factory.Name, factory);
+            _transportNameToEndpointFactory.Add(transportName, (factory, transport));
+            _transportToEndpointFactory.Add(transport, factory);
         }
 
         // Finds an endpoint factory previously registered using IceAddEndpointFactory.
-        public IEndpointFactory? IceFindEndpointFactory(string transport) =>
+        public IEndpointFactory? IceFindEndpointFactory(Transport transport) =>
             _transportToEndpointFactory.TryGetValue(transport, out IEndpointFactory? factory) ? factory : null;
 
-        // Finds an endpoint factory previously registered using IceAddEndpointFactory.
-        public IEndpointFactory? IceFindEndpointFactory(EndpointType type) =>
-            _typeToEndpointFactory.TryGetValue(type, out IEndpointFactory? factory) ? factory : null;
+        // Finds an endpoint factory previously registered using IceAddEndpointFactory, using the transport's name.
+        internal IEndpointFactory? FindEndpointFactory(string transportName, out Transport transport)
```

Comment:
We should consider using a tuple for the return and remove the out param

Reply:
This was my first idea as well, but it makes the API much more awkward to use. Try it!

Reply:
```
internal (IEndpointFactory Factory, Transport Transport)? FindEndpointFactory(string transportName)

// and use it as

if (communicator.FindEndpointFactory(transportName) is(IEndpointFactory factory, Transport transport))
{
    Endpoint endpoint = factory.Create(transport, protocol, options, oaEndpoint, endpointString);
```

Reply:
Fixed as suggested.